Add `disabled-in-tests` configuration to disable arbitrary lints in tests

smoelius opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/15600)*

Adds a catch-all configuration to disable lints in tests, so that one-offs like https://github.com/rust-lang/rust-clippy/pull/8802 are not needed.

To illustrate, the following two clippy.toml files have essentially the same effect:
```toml
allow-unwrap-in-tests = true # existing configuration
```
```toml
disabled-in-tests = ["unwrap_used"] # configuration added by this PR
```
The approach is to store a disabled lints set in `clippy_utils::diagnostics`. When a lint is in the set and `--test` is passed to rustc, diagnostics are not emitted for the lint.

Two other approaches I tried were:
- **Don't register the lint's pass.** This works but disables the other lints in the pass as well. For example, `unwrap_used` is in the `methods::Methods` pass. So not registering that pass causes `expect_used`, `should_implement_trait`, etc. to be disabled as well.
- **Don't register the lint.** E.g., filter the lint from `declared_lints::LINTS`. This works but causes rustc to to produce "unknown lint" warnings on attributes like `#[allow(unwrap_used)]`.

The present PR's approach is also much simpler than the two just described. I could share code for the other approaches if desired, though.

I expect this PR to be subject to the feature freeze.

changelog: Add `disabled-in-tests` configuration to disable arbitrary lints in tests

smoelius committed:
Add `disabled-in-tests` configuration

Undo changes to clippy_utils/src/diagnostics.rs

Remove compiled executable

Use `post_expect_filter`

Simplify determination of whether `--test` was passed

Correct condition

y21 commented:
> To illustrate, the following two clippy.toml files have essentially the same effect:

They work quite differently as currently implemented though. `allow-unwrap-in-tests` checks if it's contained within an item marked with `#[cfg(test)]` and only then avoids emitting a warning, as opposed to this which disables the lint in the entire crate (even outside of test code) when `--test` is passed.

To me this seems like the wrong thing to check for, and this type of check has also lead to some confusion in the past when combining `#[expect]` and `--all-targets`, see for example #13358 and rust-lang/rust#130021. 

For example, if I have a crate with
```rs
fn foo(opt: Option<i32>) {
  #[expect(clippy::unwrap_used)]
  opt.unwrap();
}
```
and set `disabled-in-tests = ["unwrap_used"]` and run `cargo clippy --all-targets`, then this will display an "unfulfilled lint expectation" warning, even though this isn't test code. Technically, it's not *wrong* I guess, but I'd expect the use case is to specifically only disable lints in test code but still emit it anywhere else (i.e. what `allow-unwrap-in-tests` does today), not if `--test` is passed.

smoelius commented:
I pushed some commits and changed the PR to draft in the hope of getting your feedback.

Suppose I could get the changes in this [post-expect-filter](https://github.com/rust-lang/rust/compare/master...smoelius:rust:post-expect-filter) branch merged into rustc. Do you think then that the present PR might be considered for merger into Clippy?

To be more specific, should I pursue getting that branch merged into rustc?

Please note that I added a test that very closely resembles https://github.com/rust-lang/rust-clippy/pull/15600#issuecomment-3242418546: https://github.com/rust-lang/rust-clippy/blob/a9e61900d0f9b30e0c823e165c5f9bf4cdc4ad92/tests/ui-toml/disabled_in_tests_expect/expect.rs

y21 commented on src/driver.rs:
```diff
@@ -182,6 +191,45 @@ impl rustc_driver::Callbacks for ClippyCallbacks {
 }
 
+fn build_disabled_set(sess: &Session, conf: &'static Conf) -> FxHashSet<&'static str> {
+    if sess.is_test_crate() {
+        return FxHashSet::default();
+    }
```
For my own understanding, shouldn't this condition be inverted? The idea AFAIK is to essentially ignore `disabled-in-tests` and return an empty set when *not* in a test crate, that is, we don't want to register this expect hook when not in a test crate, right?

smoelius replied:
Oof. You're right.

y21 reviewed (commented):
I don't think this needs the changes in rustc. If you want to manually fulfill an `#[expect]` without showing it (I assume that's what it's for, correct?), we already do that in a bunch of places. You can look at how [`clippy_utils::fulfill_or_allowed`](
https://github.com/rust-lang/rust-clippy/blob/e9b70454e4c9584be3b22ddabd26b741aeb06c10/clippy_utils/src/lib.rs#L1722-L1730) is implemented, it uses `cx.fulfill_expectation(..)`.

----

It would technically fix the case I'd mentioned, but checking `--test` still seems fundamentally like the wrong thing to me IMO. There won't be a false "unfulfilled expectation" warning like before, but now we won't get *any* warnings for the snippet in https://github.com/rust-lang/rust-clippy/pull/15600#issuecomment-3242418546 when removing the `#[expect]` and run with `--test`, even though it's not test code.

I still think the correct thing would be to not check for a test session but where the expression being linted actually is located. What do you think? Maybe this is something worth discussing on Zulip on what the general expectation is on how exactly this type of configuration would be used. Perhaps people do think it's fine that "disabled-in-tests" really means "disable in test sessions" as is implemented, instead of "disabled in test *code*" (what e.g. `allow-unwrap-in-tests` does).

The current behavior of `disabled-in-tests = ["unwrap_used"]` is also identical to `#![cfg_attr(test, allow(clippy::unwrap_used)]` at the crate root, right?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15600/7250a1a4b47033a52b0d426f4bca3990ac45161f..a9e61900d0f9b30e0c823e165c5f9bf4cdc4ad92)*

smoelius commented:
Zulip thread (for my own reference): https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/.60disabled-in-tests.60.20configuration.20.28PR.2015600.29

smoelius commented:
@y21 Did you have any more thoughts on this PR?

smoelius commented:
@y21 Would this PR be easier to review if it were one PR (to the Rust repository) as opposed to two PRs (one to Clippy and one to the Rust repository)?

blyxyas commented:
I'll claim this, as it's being [talked about in the kernel mailing list](https://lore.kernel.org/rust-for-linux/DECDZGMVB5DX.DGX2NMWAQOSD@nvidia.com/)

r? blyxyas

blyxyas commented:
> Do you mean another Lore thread? Thanks!

Yes, I was talking about another one, while writing that message I had several message threads and got mistaken between two of those.

The one I'm talking about is this one: https://lore.kernel.org/rust-for-linux/20251117080714.876978-1-ojeda@kernel.org/, thanks for the reminder

--

And indeed, Clippy ~~~doesn't lint doctests~~~, some lints work on doctests, but those are weird and not the norm in any way. Still, this feature is very much appreciated, and I've been thinking about it for quite some time before checking if the PR existed.

smoelius commented:
@blyxyas I squashed all of the changes down to one and rebased them onto Clippy's master branch. I also rebased the [post-expect-filter](https://github.com/rust-lang/rust/compare/master...smoelius:rust:post-expect-filter) branch onto Rust's current main branch.

Would these changes be easier to review if they were in one PR directed at Rust repo rather than two PRs directed at two different repos?

smoelius commented:
r? clippy

smoelius commented:
@flip1995 This PR has some history and requires changes to Rust to support it ([post-expect-filter](https://github.com/rust-lang/rust/compare/master...smoelius:rust:post-expect-filter)).

Please tell me if you think one PR to the Rust repo would be easier to review than one PR here and one PR there.

flip1995 commented:
I feel like this is easier to review in one PR in r-l/r. I just have to remember during the sync that I need to create a changelog entry for this.

smoelius commented:
> I feel like this is easier to review in one PR in r-l/r. I just have to remember during the sync that I need to create a changelog entry for this.

Thanks very much for reviewing, @flip1995. The PR is here: https://github.com/rust-lang/rust/pull/156396

smoelius commented:
Closing per discussion in https://github.com/rust-lang/rust/pull/156396
